[SPARK-34571][SQL][CORE] Provide a more convenient way to deprecate/remove/alternate configs

Ngone51 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

This pr proposes to add `deprecated()`/ `removed()` / `alternated()` functions to `ConfigBuilder` in order to provide more convenient ways to do these stuff.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

Currently, Spark maintains the active configs and deprecated/alternated/removed configs separately,
which is not convenient when developers want to deprecate configs and not good to maintain for the long term.

After this PR, developers can deprecate/alternate/remove a config like this:

```scala
val SOME_CONFIG =
  buildConf("spark.xxx.yyy.zzz")
    .doc(".....")")
    .version("2.3.0")
    .deprecated("3.0", "because...")
    // .removed("3.0", "because...")
    // .alternated(("3.0", "spark.aaa.bbb.ccc"))
    .booleanConf
    .createWithDefault(false)
```


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Manually tested.

Ngone51 committed:
init

Ngone51 committed:
update

Ngone51 commented:
cc @cloud-fan @HyukjinKwon @MaxGekk any idea about this?

MaxGekk reviewed (commented):
> This pr proposes to add deprecated()/ removed() / alternated() functions to ConfigBuilder in order to provide more convenient ways to do these stuff.

I think if you propose new approach, you should restore all removed configs and make them as `.removed()` then. Don't think it is good idea to maintain two approach at the same time.

> maintains the active configs and deprecated/alternated/removed configs separately,
which is not convenient when developers want to deprecate configs and not good to maintain for the long term.

IMHO, gathering all removed/deprecated configs in one place is convenient. Currently, removed configs are actually removed but you propose to keep them forever (just mark them as removed), am I right?

SparkQA commented:
**[Test build #135563 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135563/testReport)** for PR 31684 at commit [`6e375d1`](https://github.com/apache/spark/commit/6e375d1aa17ae611128d4f7457f4ac8921d30286).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40144/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40144/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40144/

SparkQA commented:
**[Test build #135563 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135563/testReport)** for PR 31684 at commit [`6e375d1`](https://github.com/apache/spark/commit/6e375d1aa17ae611128d4f7457f4ac8921d30286).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135563/

Ngone51 commented:
> I think if you propose new approach, you should restore all removed configs and make them as .removed() then. Don't think it is good idea to maintain two approach at the same time.

Yeah, I was thinking that too. But I'd like to see people's feedback before going further.

> IMHO, gathering all removed/deprecated configs in one place is convenient. 

I think it's convenient (fast) for people to know what are deprecated or removed configs now. But it's not convenient for developers to maintain. For example, to deprecate a config with the current way, a developer needs to manually add `DepracatedConfig` to the deprecated list. And if the config is going to be removed in the future, the developer has to remove the `DepracatedConfig` from the deprecated list and add `RemovedConfig` to the removed list. And this process might be problematic because these two config types are separated in two places. e.g., developers may forget to remove the `DepracatedConfig` one. While with the proposed way in this PR, the developer only needs to add `deprecated()` and change from `deprecated()` to `removed()` later.

> Currently, removed configs are actually removed but you propose to keep them forever (just mark them as removed), am I right?

Yes. But those removed configs don't disappear from Spark totally as they are still maintained in the way of `RemovedConfig`.

Ngone51 commented:
Besides, I'm thinking we may be able to extend the SQL configuration document with the new proposal. For example, we can add the current status of a conf, e.g., whether it's deprecated or removed, etc. like the `version` info does. cc @beliefer

HyukjinKwon commented on core/src/main/scala/org/apache/spark/internal/config/ConfigBuilder.scala:
```diff
@@ -205,6 +206,23 @@ private[spark] case class ConfigBuilder(key: String) {
+    this
+  }
+
+  def alternated(
```
nit but `alternated` -> `alternative`

HyukjinKwon commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -808,9 +832,21 @@ private[spark] object SparkConf extends Logging {
    * @param translation A translation function for converting old config values into new ones.
    */
-  private case class AlternateConfig(
+  private[spark] case class AlternateConfig(
```
Let's fix all Alternate -> Alternative

HyukjinKwon commented:
Can we add deprecate and alternative ones first? For removal one, I would encourage to do it in a separate PR.

cloud-fan commented:
Agree with @HyukjinKwon to do deprecation first. For the removed ones, we don't really need to keep the config entries for them anymore. We just need the names of removed configs (and the successors) to throw better error messages.

beliefer commented:
@Ngone51 Record the version when deciding deprecate config, seems good.

Ngone51 commented:
Thanks for all the feedbacks..I'll start from the depreaction and alternatives first.

Ngone51 committed:
updating

Ngone51 committed:
update

SparkQA commented:
**[Test build #135666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135666/testReport)** for PR 31684 at commit [`4a3497c`](https://github.com/apache/spark/commit/4a3497cd815c3df2463c521b62d5f575a31e4306).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40248/

SparkQA commented:
**[Test build #135666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135666/testReport)** for PR 31684 at commit [`4a3497c`](https://github.com/apache/spark/commit/4a3497cd815c3df2463c521b62d5f575a31e4306).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135666/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40248/

HyukjinKwon commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -613,31 +628,7 @@ private[spark] object SparkConf extends Logging {
       DeprecatedConfig("spark.yarn.services", "3.0.0", "Feature no longer available."),
       DeprecatedConfig("spark.executor.plugins", "3.0.0",
         "Feature replaced with new plugin API. See Monitoring documentation."),
```
```suggestion
        "Feature replaced with new plugin API. See Monitoring documentation.")
```

HyukjinKwon committed:
Update core/src/main/scala/org/apache/spark/SparkConf.scala

SparkQA commented:
**[Test build #135696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135696/testReport)** for PR 31684 at commit [`efbee39`](https://github.com/apache/spark/commit/efbee39b5392e1b28092dbc2342b3e41f46092fa).

SparkQA commented:
**[Test build #135696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135696/testReport)** for PR 31684 at commit [`efbee39`](https://github.com/apache/spark/commit/efbee39b5392e1b28092dbc2342b3e41f46092fa).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135696/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40279/
